### PR TITLE
Fade out music before change occurs

### PR DESCRIPTION
**Pull request by LennyPenny**
Fixes https://bugs.openmw.org/issues/3665
First real PR so please review :smile:

### CODE REVIEW

**Comment by scrawl** [comment]
@kcat , is this code correct? I don't think we can call updateStream in the main thread when the source is also being used in the streaming thread. Also, the begin/end update isn't necessary, is it?

**Comment by LennyPenny** [comment]
Well it didn't work without it

**Comment by kcat** [comment]
updateStream is correct. It's just updating the volume and other related playback parameters, which are fine to do asynchronously with the streaming thread (which is just making sure the buffer queue stays full).

The start/finishUpdate calls aren't strictly necessary in this case, but are kind of recommended. It helps avoid redundant internal updates for OpenAL. Though in saying that, it may be better to avoid having to do it twice per sound update (once in updateSounds and again in updateMusic).

**Comment by LennyPenny** [comment]
Okay I can confirm the working without begind/end 

Though if it's recommeded, should I just wrap all the sound updaters in a begin update?

File: apps/openmw/mwsound/soundmanagerimp.hpp [hunk 1]
```diff
@@ -114,9 +114,15 @@ namespace MWSound
...
 
         void streamMusicFull(const std::string& filename);
+        void advanceMusic(const std::string& filename);
+
         void updateSounds(float duration);
         void updateRegionSound(float duration);
         void updateWaterSound(float duration);
+        void updateMusic(float duration);
+
+        float mMusicFader;
```
Comment (scrawl) [inline apps/openmw/mwsound/soundmanagerimp.hpp]: mMusicFader needs to be initialized in the constructor. Even if a variable isn't supposed to be used until it's set, it is easy to make a mistake and end up with uninitialised values somewhere. (Also, tools like coverity/cppcheck will always scream at you for members not initialized in constructor)

Actually, why is this even a member variable when it's only used as a constant?

**Comment by scrawl** [comment]
> Though if it's recommeded, should I just wrap all the sound updaters in a start/end?

Sounds good.

**Comment by kcat** [comment]
> Though if it's recommeded, should I just wrap all the sound updaters in a start/end?

The main issue there is that the update methods won't necessarily do anything on certain updates, and calling start/finishUpdate without actually changing anything can be wasteful.

A better idea would probably be to restructure the update methods so that all necessary changes can occur within ```updateSounds```'s start/finish, and avoiding it when there's nothing to update.

File: apps/openmw/mwsound/soundmanagerimp.hpp (hunk 1 above)
Comment (LennyPenny) [inline apps/openmw/mwsound/soundmanagerimp.hpp]: Oh this is a leftover from when I thought setFadeout worked differently
Will fix

**Comment by LennyPenny** [comment]
Decided to move updateMusic into the 30 times a second area because other sounds are also faded there (so no impact on listening experience) and it avoids keeping track of state when updateMusic already started an update

**Pull request merged**